File: src/tru/trus4.f90
   PROGRAM GALAHAD_TRU4_EXAMPLE  !  GALAHAD 3.3 - 25/07/2021 AT 09:15 GMT
   USE GALAHAD_TRU_double                       ! double precision version
   IMPLICIT NONE
   INTEGER, PARAMETER :: wp = KIND( 1.0D+0 )    ! set precision
   TYPE ( TRU_control_type ) :: control
   TYPE ( TRU_inform_type ) :: inform
   TYPE ( TRU_full_data_type ) :: data
   TYPE ( NLPT_userdata_type ) :: userdata
   REAL ( KIND = wp ) :: f
   INTEGER, ALLOCATABLE, DIMENSION( : ) :: H_row, H_col, H_ptr
   REAL ( KIND = wp ), ALLOCATABLE, DIMENSION( : ) :: X, G, U, V
   EXTERNAL :: FUN, GRAD, HESS
   INTEGER :: s, status, eval_status, ne
   INTEGER, PARAMETER :: n = 3
   REAL ( KIND = wp ), PARAMETER :: p = 4.0_wp
! start problem data
   ALLOCATE( X( n ), G( n ), U( n ), V( n ) )
   ALLOCATE( H_row( 0 ), H_col( 0 ), H_ptr( 0 ) )
   X = 1.0_wp                               ! start from one
! problem data complete   
   CALL TRU_initialize( data, control, inform ) ! Initialize control parameters
   control%hessian_available = .FALSE.          ! Hessian products will be used
!  control%print_level = 5
!  control%psls_control%preconditioner = - 3    ! Apply uesr's preconditioner
   CALL TRU_import( control, data, status, n,                                  &
                    'absent', ne, H_row, H_col, H_ptr )
   status = 1                                   ! Set for initial entry
   DO                                           ! Loop to solve problem
     CALL TRU_solve_reverse_without_h( data, status, eval_status,              &
                                       X, f, G, U, V )
     SELECT CASE ( status )                     ! reverse communication
     CASE ( 2 )                                 ! Obtain the objective function
       f = ( X( 1 ) + X( 3 ) + p ) ** 2 + ( X( 2 ) + X( 3 ) ) ** 2             &
              + COS( X( 1 ) )
       eval_status = 0                     ! record successful evaluation
     CASE ( 3 )                                 ! Obtain the gradient
       G( 1 ) = 2.0_wp * ( X( 1 ) + X( 3 ) + p ) - SIN( X( 1 ) )
       G( 2 ) = 2.0_wp * ( X( 2 ) + X( 3 ) )
       G( 3 ) = 2.0_wp * ( X( 1 ) + X( 3 ) + p ) + 2.0_wp * ( X( 2 ) + X( 3 ) )
       eval_status = 0                     ! record successful evaluation
     CASE ( 5 )                            ! Obtain Hessian-vector product
       U( 1 ) = U( 1 ) + 2.0_wp * ( V( 1 ) + V( 3 ) ) - COS( X( 1 ) ) * V( 1 )
       U( 2 ) = U( 2 ) + 2.0_wp * ( V( 2 ) + V( 3 ) )
       U( 3 ) = U( 3 ) + 2.0_wp * ( V( 1 ) + V( 2 ) + 2.0_wp * V( 3 ) )
       eval_status = 0                     ! record successful evaluation
     CASE ( 6 )                            ! Apply the preconditioner
       U( 1 ) = 0.5_wp * V( 1 )
       U( 2 ) = 0.5_wp * V( 2 )
       U( 3 ) = 0.25_wp * V( 3 )
       eval_status = 0                     ! record successful evaluation
     CASE DEFAULT                          ! Terminal exit from loop
       EXIT
     END SELECT
   END DO
   CALL TRU_information( data, inform, status )
   IF ( inform%status == 0 ) THEN          ! Successful return
     WRITE( 6, "( ' TRU: ', I0, ' iterations -',                               &
    &     ' optimal objective value =',                                        &
    &       ES12.4, ', ||g|| =', ES11.4, /, ' Optimal solution = ', ( 5ES12.4 ) )" )                &
     inform%iter, inform%obj, inform%norm_g, X
   ELSE                                    ! Error returns
     WRITE( 6, "( ' TRU_solve exit status = ', I6 ) " ) inform%status
   END IF
   DEALLOCATE( X, G, U, V, H_row, H_col, H_ptr )
   CALL TRU_terminate( data, control, inform )  ! Delete internal workspace
   END PROGRAM GALAHAD_TRU4_EXAMPLE

File: src/tru/C/trus4.c
/* trus4.c */
/* Spec test for the TRU C interface */

#include <stdio.h>
#include <math.h>
#include "galahad_precision.h"
#include "galahad_tru.h"

int main(void) {

    // Derived types
    void *data;
    struct tru_control_type control;
    struct tru_inform_type inform;

    // Initialize TRU
    ipc_ status;
    tru_initialize( &data, &control, &status );

    // Set user-defined control options
    control.f_indexing = false; // C sparse matrix indexing (default)
    //control.print_level = 1;

    // Set problem data
    ipc_ n = 3; // dimension
    ipc_ ne = 5; // Hesssian elements
    rpc_ x[] = {1.,1.,1.}; // start from one
    char H_type[] = "absent"; // specify Hessian-vector products

    // Reverse-communication input/output
    ipc_ eval_status;
    rpc_ f = 0.0;
    rpc_ g[n];
    rpc_ u[n], v[n];

    //control.maxit=2;
    //control.print_level=5;
    // Set Hessian storage format, structure and problem bounds
    tru_import( &control, &data, &status, n, H_type, ne, NULL, NULL, NULL );

    // Solve the problem
    while(true){ // reverse-communication loop

        // Call TRU_solve
        tru_solve_reverse_without_mat( &data, &status, &eval_status,
                                       n, x, f, g, u, v );

        // Evaluate f(x) and its derivatives as required
        if(status == 0){ // successful termination
            break; // break while loop
        }else if(status == 2){ // obtain the objective function
            f = pow(x[0] + x[2] + 4.0, 2) + pow(x[1] + x[2], 2) + cos(x[0]);
            eval_status = 0; // record successful evaluation
        }else if(status == 3){ // obtain the gradient
            g[0] = 2.0 * ( x[0] + x[2] + 4.0 ) - sin(x[0]);
            g[1] = 2.0 * ( x[1] + x[2] );
            g[2] = 2.0 * ( x[0] + x[2] + 4.0 ) + 2.0 * ( x[1] + x[2] );
            eval_status = 0; // record successful evaluation
        }else if(status == 5){ // obtain Hessian-vector product
            u[0] = u[0] + 2.0 * ( v[0] + v[2] ) - cos( x[0] ) * v[0];
            u[1] = u[1] + 2.0 * ( v[1] + v[2] );
            u[2] = u[2] + 2.0 * ( v[0] + v[1] + 2.0 * v[2] );
            eval_status = 0; // record successful evaluation
        }else if(status == 6){ // apply the preconditioner
            u[0] = 0.5 * v[0];
            u[1] = 0.5 * v[1];
            u[2] = 0.25 * v[2];
            eval_status = 0; // record successful evaluation
        }else{ // an error has occured
            break; // break while loop
        }
    }

    // Record solution information
    tru_information( &data, &inform, &status );

    // Print solution details
    printf("iter: %" d_ipc_ " \n", inform.iter);
    printf("x: ");
#ifdef REAL_128
    for(ipc_ i = 0; i < n; i++) printf("%f ", (double)x[i]);
    printf("\nobjective: %f \ngradient: ", (double)inform.obj);
    for(ipc_ i = 0; i < n; i++) printf("%f ", (double)g[i]);
    printf("\nf_eval: %" d_ipc_ " \n", inform.f_eval);
    printf("time: %f \n", (double)inform.time.clock_total);
#else
    for(ipc_ i = 0; i < n; i++) printf("%f ", x[i]);
    printf("\nobjective: %f \ngradient: ", inform.obj);
    for(ipc_ i = 0; i < n; i++) printf("%f ", g[i]);
    printf("\nf_eval: %" d_ipc_ " \n", inform.f_eval);
    printf("time: %f \n", inform.time.clock_total);
#endif
    printf("status: %" d_ipc_ " \n", inform.status);

    // Delete internal workspace
    tru_terminate( &data, &control, &inform );

    return 0;
}
